### PR TITLE
[Forms] Further improve Xamarin.Forms support in .NET Standard projects

### PR DESCRIPTION
**Pull request by mrward**
Various .NET Standard - Xamarin.Forms problems addressed:

 - Add ContentPage with xaml then change build action to None for .xaml file.
 - Add ContentPage with xaml. Then both .xaml and .xaml.cs file removed from the project but not deleted. Then edit .csproj and remove associated MSBuild items
 - Add ContentPage with xaml then rename .xaml file
 - Add ContentPage with xaml then rename .xaml.cs file
 - Moving/copying xaml files from existing PCL/shared projects to .NET Standard project

VSTS 505912

### CODE REVIEW

**Comment by mrward** [comment]
Fixed one test failure which was affecting both VSTS and Wrench. On Wrench there is still one failing Xamarin.Forms test, which is a new test added for this new feature branch, due to some odd NuGet restore problem.

```
  Building solution configuration "Debug|Any CPU".
/Library/Frameworks/Mono.framework/Versions/5.10.0/lib/mono/msbuild/15.0/bin/NuGet.targets(103,5): error : '5d601583ce45875169e2639e723096e917887369' is not a valid version string. [md-addins/Xamarin.Forms.Addin.Tests/tests/tmp/NetStandardXamarinForms.sln-1/NetStandardXamarinForms.sln]
/Library/Frameworks/Mono.framework/Versions/5.10.0/lib/mono/msbuild/15.0/bin/NuGet.targets(103,5): error : Parameter name: value [md-addins/Xamarin.Forms.Addin.Tests/tests/tmp/NetStandardXamarinForms.sln-1/NetStandardXamarinForms.sln]
Done Building Project "/Users/builder/data/lanes/5931/5d601583/source/md-addins/Xamarin.Forms.Addin.Tests/tests/tmp/NetStandardXamarinForms.sln-1/NetStandardXamarinForms.sln" (Restore target(s)) -- FAILED.
```

This test works OK on VSTS. I cannot reproduce this failure when running the tests locally on Mono 5.8 nor Mono 5.10.

Going to try running msbuild with diagnostic logging to see if that provides more information.

**Comment by mrward** [comment]
Looks like Wrench sets some extra environment variables which are causing problems for NuGet restore.

VERSION = e518f1be030f6502d3159617d660e31183d927ec
PackageVersion = e518f1be030f6502d3159617d660e31183d927ec

Either the Version or the PackageVersion environment variable I suspect is causing NuGet to try and parse an invalid version number. I would guess PackageVersion is taken from the Version environment variable. Should be able to fix this by just clearing the version environment variable in the test.

**Comment by mrward** [comment]
Wrench build is now green.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs [hunk 1]
```diff
@@ -956,6 +956,13 @@ public string[] GetBuildActions ()
 			if (buildActions != null)
 				return buildActions;
 
+			Predicate<string> predicate = null;
+			buildActions = GetBuildActions (predicate);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs]: Style, minor, replace these two lines with:
`buildActions = GetBuildActions (predicate: null)`

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs [hunk 2]
```diff
@@ -489,11 +489,14 @@ static void RemoveItem (ProjectInfo project, MSBuildItem item, string remove, bo
 		static void AddRemoveToGlobInclude (ProjectInfo project, MSBuildItem item, string remove)
 		{
 			var exclude = ExcludeToRegex (remove);
-			foreach (var globInclude in project.GlobIncludes.Where (g => g.Item.Name == item.Name)) {
-				if (globInclude.RemoveRegex != null)
-					exclude = globInclude.RemoveRegex + "|" + exclude;
-				globInclude.RemoveRegex = new Regex (exclude);
-			}
+			do {
+				foreach (var globInclude in project.GlobIncludes.Where (g => g.Item.Name == item.Name)) {
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs]: Is there something we can cache here? We might be able to not recompute all the remove regexes if we cache the parent's somewhere.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs [hunk 3]
```diff
@@ -3640,6 +3724,25 @@ IEnumerable<MSBuildItem> FindUpdateItemsForItem (MSBuildItem globItem, string in
 			}
 		}
 
+		bool UpdateGlobHasMatchingPropertyValue (MSBuildProperty p, IMSBuildItemEvaluated evalItem)
+		{
+			MSBuildEvaluationContext context = null;
+
+			foreach (var updateItem in evalItem.SourceItems.Where (it => it.IsUpdate)) {
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs]: `UpdateGlobHasMatchingPropertyValue` is called in a loop. Please unwrap the LINQ into the foreach, so we don't allocate enumerators.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/ProjectFileNodeBuilder.cs [hunk 4]
```diff
@@ -193,6 +190,60 @@ public override void OnRenameStarting (ref int selectionStart, ref int selection
...
+		{
+			if (!file.HasChildren)
+				return null;
+
+			List<Tuple<ProjectFile, string>> files = null;
+
+			string oldName = file.FilePath.FileName;
+			foreach (ProjectFile child in file.DependentChildren) {
+				string oldChildName = child.FilePath.FileName;
+				if (oldChildName.StartsWith (oldName)) {
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/ProjectFileNodeBuilder.cs]: Use a `StringComparison` here

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/ProjectFileNodeBuilder.cs [hunk 5]
```diff
@@ -193,6 +190,60 @@ public override void OnRenameStarting (ref int selectionStart, ref int selection
...
+					string childNewName = newName + oldChildName.Substring (oldName.Length);
+
+					if (files == null)
+						files = new List<Tuple<ProjectFile, string>> ();
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/ProjectFileNodeBuilder.cs]: `ValueTuple`?

**Comment by mrward** [comment]
Updated the code. Not looked at the caching suggestion - there is similar code in other places so I think it can be looked at later on.

**Pull request merged**